src/throttle.rs

```diff
@@ -0,0 +1,19 @@
+use async_std::sync::Sender;
+use tokio::time;
+
+use crate::GooseConfiguration;
+
+/// Throttle thread, limits how fast requests can be made across all GooseUser threads.
+pub async fn throttle_main(configuration: GooseConfiguration, throttle_sender: Sender<bool>) {
+    let sleep_duration =
+        time::Duration::from_secs_f32(1.0 / configuration.throttle_requests.unwrap() as f32);
+    info!("throttle allowing 1 request every {:?}", sleep_duration);
+
+    // Loop until parent thread exits, adding one element
+    loop {
+        time::delay_for(sleep_duration).await;
```

Comment:
All looks good - except this part.

A Token Bucket Filter is waiting exactly one second and then putting k tokens until the bucket is full.

Right now if no one is consuming tokens the req/s can be way higher than configured.

Maybe a bounded channel with try_send would work?

Reply:
We create a bounded channel here:

```rust
             let (sender, receiver): (Sender<bool>, Receiver<bool>) = channel(throttle_requests);
```

Specifically `throttle_requests` is the value passed in with the `--throttle-requests` flag.

This means if nobody consumes tokens for some time, the bucket will just fill up to this maximum value then `throttle_sender.send(true).await` will block.

The only reason I added a comment about `try_send` is when we auto-tune we'll need to handle a full bucket (ie, increase the number of user threads).

Reply:
Also, we space out the adding of tokens as opposed to filling the bucket every second to avoid bursty traffic.

Reply:
Ah - I indeed missed the bounded channel argument.

So that part is good.

I wonder about the clock resolution of the wait.

"
Implementers of this algorithm on platforms lacking the clock resolution necessary to add a single token to the bucket every 
1
/
r
1/r seconds may want to consider an alternative formulation. Given the ability to update the token bucket every S milliseconds, the number of tokens to add every S milliseconds = 
(
r
∗
S
)
/
1000
(r*S)/1000.
„

So likely to instead add tokens every 10 ms might be a good compromise.

Reply:
I've added logic to remove multiple tokens if necessary to keep the sleep >= 10ms. In theory we can support throttled rates up to 1,000,000 requests per second, but likely additionally optimizations would be necessary to truly operate at that rate.